Tests/ApolloWebsocketTests/StarWarsSubscriptionTests.swift

```diff
@@ -252,4 +254,107 @@ class StarWarsSubscriptionTests: XCTestCase {
     subJedi.cancel()
     subNewHope.cancel()
   }
+  
+  // MARK: Data races tests
+  
+  func testConcurrentSubscribing() {
+    let firstSubscription = ReviewAddedSubscription(episode: .empire)
+    let secondSubscription = ReviewAddedSubscription(episode: .empire)
+    
+    let expectation = self.expectation(description: "Subscribers connected and received events")
+    expectation.expectedFulfillmentCount = 2
+    
+    var sub1: Cancellable?
+    var sub2: Cancellable?
+    
+    let queue = DispatchQueue(label: "com.apollographql.testing", attributes: .concurrent)
+    
+    queue.async {
+      sub1 = self.client.subscribe(subscription: firstSubscription) { _ in
+        expectation.fulfill()
+      }
+    }
+    
+    queue.async {
+      sub2 = self.client.subscribe(subscription: secondSubscription) { _ in
+        expectation.fulfill()
+      }
+    }
+    
+    // dispatched with a barrier flag to make sure
+    // this is performed after subscription calls
+    queue.sync(flags: .barrier) {
+      // dispatched on the processing queue to make sure
+      // this is performed after subscribers are processed
+      self.webSocketTransport.websocket.callbackQueue.async {
+        _ = self.client.perform(mutation: CreateReviewForEpisodeMutation(episode: .empire, review: ReviewInput(stars: 5, commentary: "The greatest movie ever!")))
+      }
+    }
+    
+    waitForExpectations(timeout: 10, handler: nil)
+    sub1?.cancel()
+    sub2?.cancel()
+  }
+  
+  func testConcurrentSubscriptionCancellations() {
+    let firstSubscription = ReviewAddedSubscription(episode: .empire)
+    let secondSubscription = ReviewAddedSubscription(episode: .empire)
+    
+    let expectation = self.expectation(description: "Subscriptions cancelled")
+    expectation.expectedFulfillmentCount = 2
+    
+    let sub1 = client.subscribe(subscription: firstSubscription) { _ in }
+    let sub2 = client.subscribe(subscription: secondSubscription) { _ in }
+    
+    DispatchQueue.global().async {
+      sub1.cancel()
+      expectation.fulfill()
```

Comment:
You mean that you would like to test if the subscriber is actually removed from the `WebSocketTransport`? This could be done if the processing in `disconnect()` is done using `sync` instead of `async`. 

But I think we are just testing here if two subscribers can _attempt_ to cancel concurrently.
Maybe the description is flawed?

Reply:
More that `sub1` was, in fact, cancelled. But I suppose that'd require subclassing the subscription to check whether `cancel` got called

Reply:
(punting on this too)